File: hot_chocolate_small/SmallSample/entities/Building.cs
using System;
using System.Collections.Generic;

using HotChocolate;

using x10.hotchoc.SmallSample.Repositories;

namespace x10.hotchoc.SmallSample.Entities {
  // Enums
  public enum MailboxTypeEnum {
    InBuilding,
    Community,
    Individual,
  }

  public enum PetPolicyEnum {
    NoPets,
    AllPetsOk,
    CatsOnly,
    DogsOnly,
  }


  /// <summary>
  /// A physical building which contains rental units
  /// </summary>
  public class Building : Base {
    // Regular Attributes
    [GraphQLNonNullType]
    public string? Moniker { get; set; }
    [GraphQLNonNullType]
    public string? Name { get; set; }
    [GraphQLNonNullType]
    public string? Description { get; set; }
    [GraphQLNonNullType]
    public DateTime? DateOfOccupancy { get; set; }
    [GraphQLNonNullType]
    public MailboxTypeEnum? MailboxType { get; set; }
    public PetPolicyEnum? PetPolicy { get; set; }
    [GraphQLNonNullType]
    public bool MailingAddressSameAsPhysical { get; set; }

    // To String Representation
    [GraphQLNonNullType]
    public string? ToStringRepresentation {
      get { return Name; }
      set { /* Needed to make Hot Chocolate happy */ }
    }

    // Associations
    [GraphQLNonNullType]
    public List<Unit>? Units { get; set; }
    [GraphQLNonNullType]
    public Address? PhysicalAddress { get; set; }
    public Address? MailingAddress { get; set; }

    public override void EnsureUniqueDbid() {
      base.EnsureUniqueDbid();
      Units?.ForEach(x => x.EnsureUniqueDbid());
      PhysicalAddress?.EnsureUniqueDbid();
      MailingAddress?.EnsureUniqueDbid();
    }

    internal override void SetNonOwnedAssociations(IRepository repository) {
      base.SetNonOwnedAssociations(repository);

      if (Units != null)
        foreach (Unit units in Units)
          units.SetNonOwnedAssociations(repository);

      PhysicalAddress?.SetNonOwnedAssociations(repository);

      MailingAddress?.SetNonOwnedAssociations(repository);
    }
  }
}

File: platforms/ts-chakra/src/x10_generated/small/entities/Building.ts
import * as React from 'react';
import { v4 as uuid } from 'uuid';

import { addError, type FormError } from 'react_lib/form/FormProvider';
import { getYear } from 'react_lib/type_helpers/dateFunctions';
import isBlank from 'react_lib/utils/isBlank';
import toNum from 'react_lib/utils/toNum';

import { AppContext } from 'SmallAppContext';
import { addressCalculateErrors, createDefaultAddress, type Address } from 'x10_generated/small/entities/Address';
import { unitCalculateErrors, type Unit } from 'x10_generated/small/entities/Unit';

import { MailboxTypeEnum, PetPolicyEnum } from '__generated__/graphql';


// Type Definition
export type Building = {
  readonly id: string,
  readonly moniker: string,
  readonly name?: string,
  readonly description?: string,
  readonly dateOfOccupancy?: string | null | undefined,
  readonly mailboxType?: MailboxTypeEnum | null | undefined,
  readonly petPolicy?: PetPolicyEnum | null | undefined,
  readonly mailingAddressSameAsPhysical?: boolean,
  readonly units?: Unit[],
  readonly physicalAddress?: Address,
  readonly mailingAddress?: Address,
};


// Enums
export const MailboxTypeEnumPairs: {
  value: MailboxTypeEnum,
  label: string
}[] = [
  {
    value: MailboxTypeEnum.InBuilding,
    label: 'Mailroom in Building',
  },
  {
    value: MailboxTypeEnum.Community,
    label: 'Postal System Community Mailbox',
  },
  {
    value: MailboxTypeEnum.Individual,
    label: 'Mail Delivered to Unit',
  },
];

export const PetPolicyEnumPairs: {
  value: PetPolicyEnum,
  label: string
}[] = [
  {
    value: PetPolicyEnum.NoPets,
    label: 'No Pets',
  },
  {
    value: PetPolicyEnum.AllPetsOk,
    label: 'All Pets Ok',
  },
  {
    value: PetPolicyEnum.CatsOnly,
    label: 'Cats Only',
  },
  {
    value: PetPolicyEnum.DogsOnly,
    label: 'Dogs Only',
  },
];



// Derived Attribute Functions
export function buildingAgeInYears(building?: {
    dateOfOccupancy?: string | null | undefined,
} | null | undefined): number | null | undefined | undefined {
  if (building == null) return null;
  const appContext = React.useContext(AppContext);
  const result = getYear(appContext?.today) - getYear(building?.dateOfOccupancy);
  return isNaN(result) ? null : result;
}

export function buildingApplicableWhenForMailingAddress(building?: {
    mailingAddressSameAsPhysical?: boolean,
} | null | undefined): boolean | undefined {
  if (building == null) return false;
  const result = !building?.mailingAddressSameAsPhysical;
  return result;
}



// Create Default Function
export function createDefaultBuilding(): Building {
  return {
    id: uuid(),
    moniker: '1',
    name: '',
    description: '',
    dateOfOccupancy: undefined,
    mailboxType: MailboxTypeEnum.InBuilding,
    petPolicy: undefined,
    mailingAddressSameAsPhysical: true,
    units: [],
    physicalAddress: createDefaultAddress(),
    mailingAddress: createDefaultAddress(),
  };
}


// Validations
export function buildingCalculateErrors(building?: Building, prefix?: string, inListIndex?: number): FormError[] {
  const appContext = React.useContext(AppContext);
  const errors: FormError[] = [];
  if (building == null ) return errors;

  if (isBlank(building.name))
    addError(errors, 'Name is required', ['name'], prefix, inListIndex);
  if (isBlank(building.dateOfOccupancy))
    addError(errors, 'Date Of Occupancy is required', ['dateOfOccupancy'], prefix, inListIndex);
  if (isBlank(building.mailboxType))
    addError(errors, 'Mailbox Type is required', ['mailboxType'], prefix, inListIndex);

  building.units?.forEach((x, ii) => errors.push(...unitCalculateErrors(x, 'units', ii)));
  errors.push(...addressCalculateErrors(building.physicalAddress, 'physicalAddress'));
  if (buildingApplicableWhenForMailingAddress(building))
    errors.push(...addressCalculateErrors(building.mailingAddress, 'mailingAddress'));

  if (toNum(building?.dateOfOccupancy) > toNum(appContext?.today))
    addError(errors, 'Occupancy date cannot be in the future', ['dateOfOccupancy'], prefix, inListIndex);

  return errors;
}

